disk: flashdisk: init in `POST_KERNEL`, default from DT

JordanYates opened the pull request:
Initialise flashdisk devices in `POST_KERNEL` instead of `APPLICATION`. This aligns with the other disk drivers, which are all `POST_KERNEL`, and causes no problems as `disk_access_register` is a purely software action and does not perform any operations on the underlying device.

JordanYates committed:
disk: flashdisk: init in `POST_KERNEL`

Initialise flashdisk devices in `POST_KERNEL` instead of `APPLICATION`.
This aligns with the other disk drivers, which are all `POST_KERNEL`,
and causes no problems as `disk_access_register` is a purely software
action and does not perform any operations on the underlying device.

Signed-off-by: Jordan Yates <jordan@embeint.com>

JordanYates committed:
disk: flashdisk: default enable if dts exists

Enable `DISK_DRIVER_FLASH` by default if the devicetree node exists, as
the other disk drivers do.

Signed-off-by: Jordan Yates <jordan@embeint.com>

de-nordic commented on drivers/disk/Kconfig.flash:
```diff
@@ -3,6 +3,7 @@
 
 config DISK_DRIVER_FLASH
 	bool "Flash"
+	default y if DT_HAS_ZEPHYR_FLASH_DISK_ENABLED
```
This means that in, for example, MCUboot you have to explicitly disable the `CONFIG_DISK_DRIVER_FLASH`, as it will be compiled, whether used or not, just because application that is using the same overlay has the flash disk enabled.

JordanYates replied:
Only if `DISK_DRIVERS` are enabled? And if `DISK_DRIVERS` are enabled, then enabling the Flash disk driver seems like the correct behavior

de-nordic replied:
Ah, true.

jfischer-no commented on drivers/disk/Kconfig.flash:
```diff
@@ -3,6 +3,7 @@
 
 config DISK_DRIVER_FLASH
 	bool "Flash"
+	default y if DT_HAS_ZEPHYR_FLASH_DISK_ENABLED
```
```
	default y
	depends on DT_HAS_ZEPHYR_FLASH_DISK_ENABLED
```

JordanYates replied:
The other disk drivers all use the form in the PR
https://github.com/zephyrproject-rtos/zephyr/blob/f320535f84db3b9bef8c1e975aa2fe61673c95c0/drivers/disk/Kconfig.mmc#L6
https://github.com/zephyrproject-rtos/zephyr/blob/f320535f84db3b9bef8c1e975aa2fe61673c95c0/drivers/disk/Kconfig.ram#L6
https://github.com/zephyrproject-rtos/zephyr/blob/f320535f84db3b9bef8c1e975aa2fe61673c95c0/drivers/disk/Kconfig.sdmmc#L9-L10

jfischer-no replied:
This is the usual pattern, feel free to fix it in Kconfig.mmc and Kconfig.ram as well. Looks like Kconfig.sdmmc would need more cleanup.

JordanYates replied:
I have updated this, but please remember that asking for additional changes is against the reviewer expectations.
It feels bad to implement something according to the existing pattern and then be told to change the pattern.

jfischer-no replied:
Oh, was it too much to ask that you make your changes properly? https://github.com/zephyrproject-rtos/zephyr/pull/75753#discussion_r1673636391

JordanYates committed:
tests: fs: fat_fs_api: disable flash disk in RAM test

Disable the flash disk driver in the RAM driver test to prevent the
test using the wrong disk in the tests.

Signed-off-by: Jordan Yates <jordan@embeint.com>

JordanYates committed:
disk: use standard DT kconfig dependencies

Use `depends on DT_HAS_* default y` instead of `default y if DT_HAS_*`
as the driver depends on devicetree instance.

Signed-off-by: Jordan Yates <jordan@embeint.com>

aescolar commented:
Note the linked issue
